Client/U2FHelper.swift

```diff
@@ -0,0 +1,615 @@
+                guard let key = U2FRegisteredKey(obj: regKey) else { continue }
+                self.registeredKeys.append(key)
+            }
+        }
+
+        // Check for required fields
+        if (action == kActionSign) && (self.challenge == nil) {
+            return nil
+        }
+    }
+}
+
+// The delegate just facilitates UI interactions and communications
+// with content in the WKWebView
+@available(iOS 9, *)
+protocol U2FHelperDelegate: class {
+    func u2fFinish(id id: String, response: U2FResponse)
+}
+
+// OpenSSLToken uses the Sync wrapper around OpenSSL to provide the raw crypto
+// functions required for U2F, i.e., key pair generation and signing.  Before
+// performing any operation, it prompts for a TouchID authentication.
+//
+// The key handles it returns are of the following form:
+//
+//   sha256(appID) || iv || Mac(Enc(iv, privateKey)) || Enc(iv, privateKey)
+//
+// Where:
+//
+//  - appID is the appID of the site that created the key
+//  - Enc and Mac are done with symmetric keys generated by the token and stored in the keychain
+//  - privateKey is in the binary representation provided by the ECDSAPrivateKey interface
+class OpenSSLToken {
+    private func getKeys() -> KeyBundle? {
+        if KeychainWrapper.hasValueForKey(kMasterKeyName) {
+            let keyData = KeychainWrapper.stringForKey(kMasterKeyName)!
+            let split = keyData.startIndex.advancedBy(kMasterEncKeyLength)
+            let enc = keyData.substringToIndex(split)
+            let mac = keyData.substringFromIndex(split)
+            return KeyBundle(encKeyB64: enc, hmacKeyB64: mac)
+        }
+        return nil
+    }
+
+    private func getOrCreateKeys() -> KeyBundle {
+        guard let keys = getKeys() else {
+            let keys = KeyBundle.random()
+            let rawKeys = keys.asPair()
+            let totalKey = rawKeys[0] + rawKeys[1]
+            KeychainWrapper.setString(totalKey, forKey: kMasterKeyName)
+            return keys
+        }
+        return keys
+    }
+
+    private func getNextCounter() -> NSData {
+        var counter: Int32 = 0
+        if let counterData = KeychainWrapper.dataForKey(kCounterName) {
+            counterData.getBytes(&counter, length: sizeof(Int32))
+        }
+
+        counter += 1
+        let counterData = NSData(bytes: &counter, length: sizeof(Int32))
+        KeychainWrapper.setData(counterData, forKey: kCounterName)
+        return counterData
+    }
+
+    private func authenticate(result: (Bool -> ())) {
+        // We manage the TouchID prompt ourselves (instead of using, say AppAuthenticator)
+        // because we really want it to appear on every call.
+        let context = LAContext()
+        let policy = LAPolicy.DeviceOwnerAuthenticationWithBiometrics
+        if (context.canEvaluatePolicy(policy, error: nil)) {
+            context.evaluatePolicy(policy, localizedReason: kAuthenticationReason, reply:{ authenticated, error in
+                if error != nil {
```

Comment:
It'd probably be valuable for debugging to be able to log that error someplace, but I don't know how the logging works in firefox-ios.